component/newstorage/mem/mem.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	return entry.tags, nil
+}
+
+// GetBulk fetches the values associated with the given keys.
+// If no data exists under a given key, then a nil []byte is returned for that value. It is not considered an error.
+// If any of the given keys are empty, then an error will be returned.
+func (m *memStore) GetBulk(keys ...string) ([][]byte, error) {
+	if keys == nil {
+		return nil, errors.New("keys string slice cannot be nil")
+	}
+
+	for _, key := range keys {
+		if key == "" {
+			return nil, errEmptyKey
+		}
+	}
+
+	values := make([][]byte, len(keys))
+
+	m.RLock()
+	defer m.RUnlock()
+
+	for i, key := range keys {
+		values[i] = m.db[key].value
+	}
+
+	return values, nil
+}
+
+// Query returns all data that satisfies the expression. Expression format: TagName:TagValue.
+// If TagValue is not provided, then all data associated with the TagName will be returned.
+// For now, expression can only be a single tag Name + Value pair.
+// memStore does not make use of newstorage.QueryOptions.
+func (m *memStore) Query(expression string, _ ...newstorage.QueryOption) (newstorage.Iterator, error) {
+	if expression == "" {
+		return &memIterator{}, errInvalidQueryExpressionFormat
+	}
+
+	expressionSplit := strings.Split(expression, ":")
+	switch len(expressionSplit) {
+	case expressionTagNameOnlyLength:
+		expressionTagName := expressionSplit[0]
+
+		m.RLock()
+		defer m.RUnlock()
+
+		keys, dbEntries := m.getMatchingKeysAndDBEntries(expressionTagName, "")
+
+		return &memIterator{keys: keys, dbEntries: dbEntries}, nil
+	case expressionTagNameAndValueLength:
+		expressionTagName := expressionSplit[0]
+		expressionTagValue := expressionSplit[1]
+
+		m.RLock()
+		defer m.RUnlock()
+
+		keys, dbEntries := m.getMatchingKeysAndDBEntries(expressionTagName, expressionTagValue)
+
+		return &memIterator{keys: keys, dbEntries: dbEntries}, nil
+	default:
+		return &memIterator{}, errInvalidQueryExpressionFormat
+	}
+}
+
+// Delete deletes the key + value pair (and all tags) associated with key.
+// If key is empty, then an error will be returned.
+func (m *memStore) Delete(k string) error {
+	if k == "" {
+		return errEmptyKey
+	}
+
+	m.Lock()
+	defer m.Unlock()
+	delete(m.db, k)
+
+	return nil
+}
+
+// Batch performs multiple Put and/or Delete operations in order.
+// If any of the given keys are empty, then an error will be returned.
+func (m *memStore) Batch(operations []newstorage.Operation) error {
+	m.Lock()
+	defer m.Unlock()
+
+	for _, operation := range operations {
+		if operation.Key == "" {
+			return errEmptyKey
+		}
+	}
+
+	for _, operation := range operations {
+		if operation.Value == nil {
+			delete(m.db, operation.Key)
+			continue
+		}
+
+		m.db[operation.Key] = dbEntry{
+			value: operation.Value,
+			tags:  operation.Tags,
+		}
+	}
+
+	return nil
+}
+
+// Close closes this store object, freeing resources and deleting all data stored within.
+func (m *memStore) Close() error {
+	m.Lock()
+	defer m.Unlock()
+
+	m.db = make(map[string]dbEntry)
```

Comment:
Not sure if this is necessary. The garbage collector should take care of it.

Reply:
Removed the unnecessary lines.